add scripts for Fastsurfer orchestration on slurm clusters

dkuegler opened the pull request:
Includes the bias field commits

Adds three files to run Fastsurfer on a subject_list file and on a slurm cluster in a resource-optimized way.

Known issues:
On our slurm cluster I experienced some sporadic out-of-memory issues where they should not have happened. These need to be investigated.

dkuegler committed:
add ifh2hdr into docker image
fix python-s in run_fastsurfer.sh (only use this if it is defined, else use python3.10 -- this is the default value and the in-docker fix to ignore user-space packages)
fix bashrc definition in Docker/Dockerfile

dkuegler committed:
add overloads for readITKimage in image_io.py
rewrite/update N4_bias_correct.py
- accept additional parameters (--aseg, --verbosity, --version)
- port from optparse to argparse
- change printing to logging facilities (instead of print), verbosity changes amount of prints.
- rename functions and variables to follow PEP8 (lowercase, underscore to separate)
- add function and integration to process the aseg and calculate wm intensity based on the wm segmentation from aseg
- change normalization code to keep the zero-base (robust scaling should be done previously, e.g. typically in conforming)
- rearrange code structure so functions are not as dependent on flags. Specifically, whether or not specific flags/files are present is moved to the main script. Consequently, functions are more predictable and not dependent on arguments.

dkuegler committed:
run_fastsurfer
- minimal reorder and formatting

add
- brun_fastsurfer.sh: a batch version of run_fastsurfer.sh
- srun_fastsurfer.sh: a slurm/orchestration script for run_fastsurfer.sh (and brun_fastsurfer.sh)
- stool.sh: functions for srun_* and brun_*

m-reuter committed:
fix doc string talairachXFM
fix doc string talairach origin

m-reuter commented:
Nice! I wonder if it makes sense to move these into a slurm subdirectory or into the singularity subdir?

dkuegler committed:
add --slurm_jobarray argument

dkuegler committed:
fix call to N4_bias_correct in run_fastsurfer.sh

dkuegler committed:
update the help text of N4_bias_correct.py
change the renormalization (--skipwm mode) to only consider the average intensity of the masked image.
fix docstring in image_io.py

dkuegler committed:
fix check_hpc_work if no --work given (never exists)
add statusfile and do not fail in brun_fastsurfer.sh so a single segmentation failed does not break the pipeline
fix double prepend in_dir in check_subject_images
cleanup the job dependencies of seg, surf and cleanup

dkuegler committed:
add statusfile to debug output
fix wait in brun_fastsurfer
add Note/Error message for invalid path to t1 in run_fastsurfer.sh
update srun_fastsurfer docs:
- usage: output folder structure
- check of out_dir
- update of statusfile
add --no-home for singularity
fix copyjob for --surf_only (srun_fastsurfer)
add time (timelimit)
fix check_seg_surf_only()
move out_folders into a slurm folder
fix $dir in cleanup
fix copy script

dkuegler committed:
Add info to logfiles if execution is aborted (trap)
no jobstep in segmentation
fix premature creation of the hpc_work directory
create hpc_work dir on demand
make cleanup script only delete the folder if all the moving was successful
